source/common/config/utility.cc

```diff
@@ -275,5 +275,36 @@ envoy::api::v2::ClusterLoadAssignment Utility::translateClusterHosts(
   return load_assignment;
 }
 
+void Utility::translateOpaqueConfig(const ProtobufWkt::Any& typed_config,
+                                    const ProtobufWkt::Struct& config,
+                                    Protobuf::Message& out_proto) {
+  static const std::string& struct_type =
+      ProtobufWkt::Struct::default_instance().GetDescriptor()->full_name();
+
+  if (!typed_config.value().empty()) {
+
+    // Unpack methods will only use the fully qualified type name after the last '/'.
+    // https://github.com/protocolbuffers/protobuf/blob/3.6.x/src/google/protobuf/any.proto#L87
+    absl::string_view type = typed_config.type_url();
+    size_t pos = type.find_last_of('/');
+    if (pos != absl::string_view::npos) {
+      type = type.substr(pos + 1);
+    }
+
+    // out_proto is expecting Struct, unpack directly
+    if (type != struct_type || out_proto.GetDescriptor()->full_name() == struct_type) {
+      typed_config.UnpackTo(&out_proto);
+    } else {
+      ProtobufWkt::Struct struct_config;
+      typed_config.UnpackTo(&struct_config);
+      MessageUtil::jsonConvert(struct_config, out_proto);
+    }
```

Comment:
Shouldn't there be a return after this line?

Reply:
It wouldn't hurt, but I think the invariant is that !typed_config.empty ^ !config.empty, i.e. they should generally be disjoint via oneof.